platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +83,131 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+    AssetInfo sellAsset = assetService.getAssetByName(request.getSellAsset());
     GetRateResponse.Rate rate = getRateResponse.getRate();
     if (rate == null || rate.getPrice() == null) {
-      Log.error("missing 'price' in the GET /rate response");
-      throw new ServerErrorException("internal server error");
+      logErrorAndThrow("missing 'price' in the GET /rate response", ServerErrorException.class);
     }
 
     if (request.getType() != GetRateRequest.Type.INDICATIVE) {
-      if (rate.getFee() == null) {
-        Log.error("'fee' is missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+      if (Objects.requireNonNull(rate).getFee() == null) {
+        logErrorAndThrow("'fee' is missing in the GET /rate response", ServerErrorException.class);
       }
     }
 
     if (request.getType() == GetRateRequest.Type.FIRM) {
       if (rate.getId() == null || rate.getExpiresAt() == null) {
-        Log.error("'id' and/or 'expires_at' are missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+        logErrorAndThrow(
+            "'id' and/or 'expires_at' are missing in the GET /rate response",
+            ServerErrorException.class);
+      }
+    }
+
+    // sell_amount is present and positive number
+    if (!isPositiveNumber(Objects.requireNonNull(rate).getSellAmount())) {
+      logErrorAndThrow(
+          "'sell_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount is present and positive number
+    if (!isPositiveNumber(rate.getBuyAmount())) {
+      logErrorAndThrow(
+          "'buy_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is preset, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
+        logErrorAndThrow(
+            "'fee.total' is missing or not a positive number in the GET /rate response",
+            ServerErrorException.class);
+      }
+      // fee.asset is a valid asset
+      AssetInfo feeAsset = assetService.getAssetByName(fee.getAsset());
+      if (fee.getAsset() == null || feeAsset == null) {
+        logErrorAndThrow(
+            "'fee.asset' is missing or not a valid asset in the GET /rate response",
+            ServerErrorException.class);
+      }
+
+      if (fee.getAsset().equals(request.getSellAsset())) {
+        // when fee is in sell_asset,
+        // check that sell_amount is equal to price * buy_amount + (fee ?: 0)
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()))
+                .add(new BigDecimal(fee.getTotal()))
+                .setScale(sellAsset.getSignificantDecimals(), HALF_UP);
```

Comment:
We use FLOOR for total price, can we also use it here for consistency?
https://github.com/stellar/java-stellar-anchor-sdk/blob/develop/core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java#L531

Reply:
I think it comes from Bitso's request originally. Now I think we should just stick to it, and (better) declare a constant somewhere that is currently set to FLOOR. We can allow it to be configured in the future, if requested by some anchor

Reply:
`DEFAULT_ROUNDING_MODE` added to `NumberHelper` which is a temporary placeholder for future configuration.